### PR TITLE
Trace file parse, write, build, validate

### PR DESCRIPTION
**Pull request by et22**
*Description of changes:* Add trace file parse, write, validate support with builders for easy trace file creation.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: smithy-codegen-core/src/test/java/software/amazon/smithy/codegen/core/TraceFileTest.java [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+    -Test creating, parsing, validating and writing a TraceFile from scratch.
+    -Test parse and write functionality combined on a correct trace file
+    -Test parse and write functionality with nodes with single children instead of arrays of children
+    -Test parse and write functionality combined on a correct trace file without the optional definitions section
+    */
+
+    @Test
+    void assertWriteTraceFileFromScratchWorks() throws IOException, URISyntaxException {
+        /*
+        Building ArtifactMetadata - this builder uses setTimestampAsNow, but you can also use a different
```
Comment (srchase) [inline smithy-codegen-core/src/test/java/software/amazon/smithy/codegen/core/TraceFileTest.java:27]: nit: multi-line comments should get formatted like this:
```suggestion
        /**
         * Building ArtifactMetadata - this builder uses setTimestampAsNow, but you can also use a different
```

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/Definitions.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+ * </p>
+ * <p>
+ * The tags {@link Map} defines the set of tags that can be used in a {@link ShapeLink} object.
+ * Each key is the name of the tag, and each value is the description of the tag.
+ * Tags are used to provide semantics to links. Tools that evaluate trace models
+ * use these tags to inform their analysis. For example, a tag for an AWS SDK code
+ * generator could be "requestBuilder" to indicate that a class is used as a builder for a request.
+ * </p>
+ */
+public final class Definitions implements ToNode, ToSmithyBuilder<Definitions> {
```
Comment (srchase) [inline smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/Definitions.java:50]: `Definitions` seems slightly too generic of a name for this class, unless you think these will be useful beyond ShapeLink objects.

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+    /**
+     * Finds invalid types and tags and either removes them or throws an error depending on
+     * whether toThrow is true or false.
+     *
+     * @throws ExpectationNotMetException if a type or tag in shapes is not in definitions.
+     */
+    public void validateTypesAndTags() {
+        Objects.requireNonNull(shapes);
+        Objects.requireNonNull(definitions);
+        for (Map.Entry<ShapeId, List<ShapeLink>> entry : shapes.entrySet()) {
```
Comment (srchase) [inline smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java:180]: This validation is a little convoluted to follow.  If it can be broken up, that could help with readability.

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+    private String smithyTrace;
+    private ArtifactMetadata artifactMetadata;
+    private ArtifactDefinitions artifactDefinitions; //Optional
+    private Map<ShapeId, List<ShapeLink>> shapes;
+    private SourceLocation sl = new SourceLocation("");
+
+    private TraceFile(Builder builder) {
+        smithyTrace = SmithyBuilder.requiredState(SMITHY_TRACE_TEXT, builder.smithyTrace);
+        artifactMetadata = SmithyBuilder.requiredState(ARTIFACT_TEXT, builder.artifactMetadata);
+        shapes = SmithyBuilder.requiredState(SHAPES_TEXT, MapUtils.copyOf(builder.shapes));
```
Comment (kstich) [inline smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java:84]: I don't think you need this requirement check here. The `shapes` map is defaulted to an empty map and [`SmithyBuilder.requiredState` works by checking for `null`](https://github.com/awslabs/smithy/blob/master/smithy-utils/src/main/java/software/amazon/smithy/utils/SmithyBuilder.java#L43), meaning this ends up wasted.

The file must contain an entry for `shapes`, which [you're already expecting here.](https://github.com/awslabs/smithy/pull/492/files#diff-340d4054b939dec6834ee291ebe606fbR114)

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+
+    /**
+     * Parses and validates the trace file passed in as filename
+     * and instantiates smithyTrace and defintions, and fills
+     * artifactMetadata and shapes.
+     *
+     * @param filename the absolute or relative path of tracefile
+     * @throws FileNotFoundException if filename is not found for reading
+     */
+    public static TraceFile parseTraceFile(URI filename) throws FileNotFoundException {
```
Comment (kstich) [inline smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java:96]: Probably should remove this, see the previous comment about separation of concerns with file loading.

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+    }
+
+    /**
+     * Writes the TraceFile JSON ObjectNode
+     * to construct a trace file.
+     *
+     * @param fileName absolute or relative path to write the trace file
+     * @throws IOException if there is an error writing to fileName
+     */
+    public void writeTraceFile(String fileName) throws IOException {
```
Comment (kstich) [inline smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java:140]: Probably should remove this, see the previous comment about separation of concerns with file loading.

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+                .build();
+    }
+
+    /**
+     * Finds invalid types and tags and either removes them or throws an error depending on
+     * whether toThrow is true or false.
+     *
+     * @throws ExpectationNotMetException if a type or tag in shapes is not in definitions.
+     */
+    public void validateTypesAndTags() {
```
Comment (kstich) [inline smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java:177]: This method needs a little better documentation around its precondition and use case.

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+        //The optional ArtifactDefinitions must be non-null to call this method
+        Objects.requireNonNull(artifactDefinitions);
```
Comment (kstich) [inline smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java:179]: You may want to do your own null check here that throws a clearer message in an `ExpectationNotMetException` when it fails.

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+    /**
+     * Parses model and determines whether the trace file object meets the specs of the model by checking if
+     * the trace file contains all the ShapeIds in the model and the model contains all the ShapeIDs in the trace file.
+     *
+     * @param modelResourceName the model name to validate the trace file against. Model should be in the
+     *                          resources file.
+     * @throws ExpectationNotMetException if model contains a ShapeID not in TraceFile or TraceFile contains a ShapeID
+     *                                    not in model.
+     */
+    public void validateModel(String modelResourceName) {
```
Comment (kstich) [inline smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java:213]: This is another context where a similar separation of concerns comes in to play. This function would need to bloat significantly to handle all the other things that may need to happen to a `ModelAssembler` to build a valid `Model`. Instead, you can have this function just take an instance of an already built `Model`.

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+        //model contains all the shapeIds in shapes.keySet()
+        for (ShapeId id : shapes.keySet()) {
+            model.expectShape(id);
+        }
+
+        //shapes.keySet() contains all the shapeIds in model, we don't care about shapes in smithy.api namespace
+        for (Shape shape : model.toSet()) {
+            ShapeId id = shape.getId();
+            if (!id.getNamespace().equals("smithy.api") && !shapes.containsKey(id)) {
+                throw new ExpectationNotMetException("shapes does not contain" + id.toString() + " but model does", sl);
```
Comment (kstich) [inline smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java:228]: Nit: Exception messages should be full sentences, including capitalization and punctuation.

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+
+        //model contains all the shapeIds in shapes.keySet()
+        for (ShapeId id : shapes.keySet()) {
+            model.expectShape(id);
+        }
+
+        //shapes.keySet() contains all the shapeIds in model, we don't care about shapes in smithy.api namespace
+        for (Shape shape : model.toSet()) {
+            ShapeId id = shape.getId();
+            if (!id.getNamespace().equals("smithy.api") && !shapes.containsKey(id)) {
```
Comment (kstich) [inline smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java:227]: This first check should probably be replaced with [`Prelude.isPublicPreludeShape`](https://awslabs.github.io/smithy/javadoc/1.0.6/software/amazon/smithy/model/loader/Prelude.html#isPublicPreludeShape(software.amazon.smithy.model.shapes.ToShapeId)).

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+
+    /**
+     * Builder for constructing TraceFile's from scratch.
+     */
+    public static final class Builder implements SmithyBuilder<TraceFile> {
+
+        private String smithyTrace = SMITHY_TRACE_VERSION;
+        private ArtifactDefinitions artifactDefinitions;
+        private ArtifactMetadata artifactMetadata;
+        private Map<ShapeId, List<ShapeLink>> shapes = new HashMap<>();
```
Comment (kstich) [inline smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java:294]: See previous comment about `final`, `.clear()`, and `.putAll()`.

File: smithy-codegen-core/src/test/resources/software/amazon/smithy/codegen/core/definitions.json [hunk 13]
```diff
@@ -0,0 +1,10 @@
+{
+  "types": {
+    "t1": "t1val",
+    "t2": "t2val"
+  },
+  "tags": {
+    "tag1": "tag1val",
+    "tag2": "tag2val"
+  }
+}
```
Comment (kstich) [inline smithy-codegen-core/src/test/resources/software/amazon/smithy/codegen/core/definitions.json:10]: Looks like there are still some files hanging around without empty lines at the end.

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java [hunk 14]
```diff
@@ -0,0 +1,349 @@
...
+         *
+         * @param id   ShapeId
+         * @param link ShapeLink corresponding to ShapeId
+         * @return This builder.
+         */
+        public Builder addShapeLink(ShapeId id, ShapeLink link) {
+            if (!this.shapes.containsKey(id)) {
+                this.shapes.put(id, new ArrayList<>());
+            }
+            this.shapes.get(id).add(link);
```
Comment (kstich) [inline smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java:299]: You can use `.getOrDefault(id, new ArrayList<>())` here to avoid doing the pre-check by yourself.

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/ShapeLink.java [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+    private List<String> tags; //optional
+    private String file; //optional
+    private Integer line; //optional
+    private Integer column; //optional
+
+    private ShapeLink(Builder builder) {
+        type = SmithyBuilder.requiredState(TYPE_TEXT, builder.type);
+        id = SmithyBuilder.requiredState(ID_TEXT, builder.id);
+        tags = ListUtils.copyOf(builder.tags);
+        if (tags.isEmpty()) {
```
Comment (kstich) [inline smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/ShapeLink.java:53]: Lists and Maps are usually left empty instead of set back to null like this one. This prevents needing to route through an `Optional` (like you do below), and requires only a little refactoring in your `toNode()`:

```java
// Store the mostly-complete builder
ObjectNode.Builder nodeBuilder = ObjectNode.objectNodeBuilder()
        .withMember(ID_TEXT, id)
        ...

// Apply the tags member if you have any.
if (!tags.isEmpty()) {
    nodeBuilder.withMember(TAGS_TEXT, Node.fromStrings(tags));
}

// Return the built node
return nodeBuilder.build();
```

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java [hunk 16]
```diff
@@ -0,0 +1,349 @@
...
+     *
+     * @return ObjectNode representation of a TraceFile.
+     */
+    @Override
+    public ObjectNode toNode() {
+        //constructing shapes ObjectNode map
+        ObjectNode.Builder shapesBuilder = ObjectNode.objectNodeBuilder();
+        for (Map.Entry<ShapeId, List<ShapeLink>> entry : shapes.entrySet()) {
+            shapesBuilder.withMember(entry.getKey().toString(),
+                    entry.getValue().stream().map(ShapeLink::toNode).collect(ArrayNode.collect()));
```
Comment (kstich) [inline smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java:112]: Minor: This value generation is really long, consider storing it in a variable to make it easier to read.

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java [hunk 17]
```diff
@@ -0,0 +1,349 @@
...
+     * @throws ExpectationNotMetException if model contains a ShapeID not in TraceFile or TraceFile contains a ShapeID
+     *                                    not in model.
+     */
+    public void validateModel(Model model) {
+        //model contains all the shapeIds in shapes.keySet()
+        for (ShapeId id : shapes.keySet()) {
+            model.expectShape(id);
+        }
+        //shapes.keySet() contains all the shapeIds in model, we don't care about shapes in smithy.api namespace
+        for (Shape shape : model.toSet()) {
```
Comment (kstich) [inline smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java:178]: The usability of this validation can be improved a bit by waiting until the end to throw an error containing any shape that's not present. You'd want to store any missing shape in a list, then throw if that list isn't empty. This will require a little reconfiguration of the exception message to handle it as well

File: CHANGELOG.md [hunk 18]
```diff
@@ -119,8 +119,8 @@ from Smithy `Node`s. ([#301](https://github.com/awslabs/smithy/pull/301))
 * Individual authentication schemes are now defined as individual traits that are annotated with
 [the `authDefinition` trait.](https://awslabs.github.io/smithy/1.0/spec/core/auth-traits.html#authdefinition-trait) [BC] ([#273](https://github.com/awslabs/smithy/pull/273), [#280](https://github.com/awslabs/smithy/pull/280))
   * Previously listed [authentication schemes now have trait implementations.](https://awslabs.github.io/smithy/1.0/spec/core/auth-traits.html)
-* The `smithy.api#enum` trait is now a list of enum definitions instead of a map of string keys to
-enum definitions to improve clarity and encourage adding more properties to definitions. [BC] ([#326](https://github.com/awslabs/smithy/pull/326))
+* The `smithy.api#enum` trait is now a list of enum Definitions instead of a map of string keys to
```
Comment (srchase) [inline CHANGELOG.md:122]: Was this change intentional?

File: smithy-model/src/test/java/software/amazon/smithy/model/node/NodeMapperTest.java [hunk 19]
```diff
@@ -411,6 +411,31 @@ public String getFoo() {
         }
     }
 
+    @Test
+    public void canDisableFromNodeInsideOfClass() {
+        ObjectNode objectNode = ObjectNode.objectNodeBuilder().withMember("foo", "baz").build();
+        NodeMapper mapper = new NodeMapper();
+        mapper.disableFromNodeForClass(DisabledFromNode.class);
+        DisabledFromNode result = mapper.deserialize(objectNode,DisabledFromNode.class);
```
Comment (mtdowling) [inline smithy-model/src/test/java/software/amazon/smithy/model/node/NodeMapperTest.java:419]: missing space after comma

File: smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/ArtifactDefinitions.java [hunk 20]
```diff
@@ -0,0 +1,170 @@
...
+     *
+     * @param value ObjectNode that contains the JSON data inside the definitions tag of
+     *              the trace file
+     */
+    public static ArtifactDefinitions fromNode(Node value) {
+        NodeMapper mapper = new NodeMapper();
+        mapper.disableFromNodeForClass(ArtifactDefinitions.class);
+        return mapper.deserialize(value, ArtifactDefinitions.class);
+    }
+
```
Comment (mtdowling) [inline smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/ArtifactDefinitions.java:61]: extra newline

File: smithy-model/src/main/java/software/amazon/smithy/model/node/NodeMapper.java [hunk 21]
```diff
@@ -164,15 +164,17 @@ private static String createMessage(String property, String pointer, Class<?> in
          *
          * @param nodeType Node type being converted.
          * @param target The class to create from the Node.
+         * @param nodeMapper The NodeMapper being used to call the ObjectCreator.
```
Comment (srchase) [inline smithy-model/src/main/java/software/amazon/smithy/model/node/NodeMapper.java:167]: Could you move the changes to NodeMapper in smithy-model into a separate PR?

It'd be helpful to have those split out in the commit history?
